### ISSUE TITLE
Zebra problem, chimera, supersling

### ISSUE THREAD

**Issue by jmgs23** [wstyres/Zebra#888]
**Describe the bug**

I can’t install nothing 

**To Reproduce**

**Expected behavior**

**Screenshots**
![5870889F-0880-4A29-8649-0D14267D0740](https://user-images.githubusercontent.com/61391244/75120571-11497800-5663-11ea-87a8-9a63a2bd63cc.jpeg)

**Device Information (please complete the following information):**
 - iOS Version [e.g. 12.1.2 or 10.3.3]:  12.1.4
 - Zebra Version [e.g. 1.0\~beta1 or beta1]:  latest
 - Device Type [e.g. iPhone or iPad]:  iPhone 8+
 - Jailbreak Type [e.g. unc0ver, Chimera, checkra1n, etc.]:  chimera

**Crash Log or System Log**
Please provide a Crash Log to help us solve the issue.

**Additional context**

**Comment by blanxd** [comment]
I'm encountering the same on Chimera

- iOS Version: 12.4
- Zebra Version: 1.1~beta1-1
- Device Type: iPhone 6s
- Jailbreak Type: Chimera

```
# ls -la /usr/libexec/zebra/supersling
-rwsr-sr-x  1 root  wheel  199248 Feb 22 05:45 /usr/libexec/zebra/supersling
```
Additional:
This issue does Not occur on Electra (iPhone 5s / 11.4.1 / Zebra 1.1~beta1-1), works just fine there.

( @jmgs23 may I suggest putting `supersling` and maybe `Chimera` in the title here, or something a bit more informative so it's easier to find, I was just about to open a new issue about the same thing... )

**Comment by m1ten** [comment]
What happens when you go to that directory?

**Comment by blanxd** [comment]
uum, what do you mean, @m1ten ?
```
# cd /usr/libexec/zebra/
# ls -la
total 408
drwxr-xr-x    5 root  wheel     160 Feb 27 06:27 .
drwxr-xr-x  197 root  wheel    6304 Feb 27 06:07 ..
-rwsr-sr-x    1 root  wheel    3726 Feb 22 05:44 firmware.sh
-rwsr-sr-x    1 root  wheel     535 Feb 22 05:44 startup
-rwsr-sr-x    1 root  wheel  199248 Feb 22 05:45 supersling
# ./supersling ls
# echo $?
77
#
```
Looks all cool to me here. (I don't know about any newer code in supersling of course, doesn't say CHAOS any more but ret 77 looks correct to me judging by the published code, if running it from shell like this)
Uuh yeah, my humble conclusion is it's exactly something new in supersling, because if I copy the v1.0.7 binary there then it works.

**Comment by wstyres** [comment]
I don't see why this would be affecting the Chimera jailbreak exclusively because it is mostly the same under-the-hood as the Electra jailbreaks. I don't have a device on either so my ability to debug is slim. Have you tried reinstalling Zebra or running the command `chmod 6755 /usr/libexec/zebra/supersling`?

**Comment by blanxd** [comment]
oh ok, perfectly understandable, I keep my 2 older devices on 11/12 on electra/chimera precisely for testing purposes.

Yes I reinstalled 1.0.7 and it worked (copied the supersling binary at this point), then reinstalled 1.1~beta1-1 and doesn't work again. Rebooted also, without success. Chmod'ing doesn't make a difference, it's all correct from the start as you can see from my `ls -la`.

In the logs I don't see much about this procedure, there's a lot about downloading the pkgs (TLS etc), then `Completed Downloads ( this and that )`, then next is already about uicache/respring/etc depending on the pkgs, so the error part goes silently.

I can see you have a couple new exported functions in supersling, and by some names, am I correct assuming you might be dlopening supersling instead of using NSTask? ...or something of that sort, humble assumptions only :)
I can kinda recall maybe having to treat chimera special in the past yr or 2 (or yeah could have been the diff between u0/cr vs. chimera/electra) but I cannot find it right now in my notes... (I'm using a setuid helper in a tweak, but it's all popen and posix_spawn there). 

So now I made a quick test shell binary which calls the supersling (popen and posix_spawn), I put it in place of Zebra to supposedly get past your proc_pidpath() check. And it does return EX_NOPERM, and does so also on electra if called from shell like this. The supersling from 1.0.7 returns a successful 0, so I guess you're doing some more validation in there now, but it somehow fails on Chimera. What else could I try in order to debug this?

**Comment by wstyres** [comment]
I think I'll have to go back to the drawing board on this one. If you're experiencing this issue I would recommend downgrading back to 1.0.7 since you won't be able to upgrade to a future 1.1 beta with this issue anyway.

I can give you a Zebra deb that would enable some debug messages and you can send them back to me so that I can look a little closer at the issue but I think the best course is probably just to move forward with my plans for replacing su/sling.

The extra code added in the 1.1 update was looking at the actual first launch argument to get the specific path of the application (as opposed to the symlinked path on stashed devices which can't be trusted). You can examine the code yourself [here](https://github.com/wstyres/Zebra/blob/master/Supersling/main.c) as it is open source. Let me know if you want the debug deb, I can post a link here for you to download.

**Comment by wstyres** [comment]
https://www.dropbox.com/s/x8gfll8on90lu2m/xyz.willy.zebra_1.1~beta1-1-6%2Bdebug_iphoneos-arm.deb?dl=0 Here is a debug deb that should print out some extra information when installing a package from within Zebra.

**Comment by blanxd** [comment]
Allright, the additional message with the debug deb is `Didn't get a first response from launchd`, but I suppose this doesnt't help much on this topic...?

![IMG_6432](https://user-images.githubusercontent.com/2082849/75595847-afd83d80-5a96-11ea-99dc-a742e6fc438c.JPG)

Owh, I didn't check the git history earlier, and was assuming things simply by looking at an `nm -u supersling` output, there's stuff a few commits back that look a bit more like the current symbol table in the beta 1&2 packages (compiling it from the current source works fine on chimera). Yeah finding the _real_ path behind any stashing would sure be the right thing to do, no doubt.
Once you get there, I can help test things on electra/chimera.

**Comment by wstyres** [comment]
I didn’t even link the right file, my bad. The correct file is on the 1.1 branch. 

Yeah that error message isn’t very good, that is pretty much the first check that it runs. I wonder if there are entitlements missing?

**Comment by blanxd** [comment]
oh ok :) now I found the functions, ok.
Entitlements should be just fine, the 3 necessary ones are present. It doesn't print anything but returns 77 so it's the `return` in the end there.
Will get back after playing around a bit with this (it's 3am here so probably after some hours of rest).

**Comment by pat97425** [comment]
Bonjour, j'ai installé Zebra mais je n'arrive pas à installer des paquets: j'ai un message qui dit "n'a pas pu terminer cette commande ...n'a pas les permissions situéesdans usr/libexec/zebra/supersling" et reporter sur Github.

problème vu sur le topic mais je ne comprends pas les réponses en anglais technique.

**Comment by wstyres** [comment]
@pat97425 Nous travaillons pour résoudre le problème. Il sera disponible dans la prochaine version.

**Comment by blanxd** [comment]
ok so: `launch_data_get_type(message)` resolves to `LAUNCH_DATA_ERRNO` (at Line92). `launch_data_get_errno(message)` resolves to `154`, which is "Unknown" to regular strerror(). Forgive me my ignorance, I don't know what to test next, I haven't used this api before, does it suggest launchd is being messed with differently on chimera or what... not sure if similar validation could be achieved with the xpc_... functions?

**Comment by wstyres** [comment]
@blanxd Interesting, can you install `ldid` and run `ldid -e /usr/libexec/zebra/supersling`?

**Comment by blanxd** [comment]
Yeah, the 3 ones are all there, for all 3 archs (it doesn't get stripped or anything, if that's what you wanted to make sure of), I checked it with jtool before I even started posting here. If I don't embed the ents and sign the binary with like [this cert](https://github.com/pwn20wndstuff/Undecimus/blob/master/apt/signing-certificate_0.0.1_iphoneos-arm.deb), then it doesn't error with the 154 but then of course we get Sandboxed and nothing happens at all (Killed: 9).

What I also did was scan everything in /bin /sbin /usr/bin /usr/sbin (I love [dsdump](https://github.com/DerekSelander/dsdump)), and there are only 2 binaries which use the launch_msg() function at all, notifyd and syslogd, they're both running as root, and neither of them has any entitlements embedded at all, but they're of course signed by apple. But there are lots of things using the newer [xpc api](https://developer.apple.com/documentation/xpc?language=objc) instead, also things that run as mobile and which have lots of different entitlements embedded. I kinda feel querying launchd through those, about the parent process, should be doable. These aren't deprecated like the launch_... stuff, but I have no idea if it would help in the current situation at all.

Then again, I'm thinking, using the sysctl api instead, one should also be able to find the vnodes and stuff for a process... I haven't used it precisely for this sort of procedure so far but it could be worth looking into?

**Comment by wstyres** [comment]
@blanxd Interesting, can I see the output of the command?

I also have a partially working Zebra implemented with my new solution. I've already confirmed that it works on unc0ver and checkra1n so I still need confirmation of Chimera and Electra as well as some primary beta testing before I release to the public. If you're interested let me know and I can give you the deb.

**Comment by blanxd** [comment]
uum, sry I didn't quite get which command do you mean?

But yeah I can test if it works on electra&chimera for sure, like whether there aren't any fundamental roadblocks.

**Comment by wstyres** [comment]
The `ldid -e /usr/libexec/zebra/supersling` command

**Comment by blanxd** [comment]
```
# ldid -e /usr/libexec/zebra/supersling
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
	<dict>
		<key>platform-application</key>
		<true/>
		<key>com.apple.private.skip-library-validation</key>
		<true/>
		<key>com.apple.private.security.no-container</key>
		<true/>
	</dict>
</plist>
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
	<dict>
		<key>platform-application</key>
		<true/>
		<key>com.apple.private.skip-library-validation</key>
		<true/>
		<key>com.apple.private.security.no-container</key>
		<true/>
	</dict>
</plist>
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
	<dict>
		<key>platform-application</key>
		<true/>
		<key>com.apple.private.skip-library-validation</key>
		<true/>
		<key>com.apple.private.security.no-container</key>
		<true/>
	</dict>
</plist>
```

**Comment by wstyres** [comment]
Should be fixed as of 22a43912cb117b38789daf211d32028af107e548

**Comment by blanxd** [comment]
> Should be fixed as of [22a4391](https://github.com/wstyres/Zebra/commit/22a43912cb117b38789daf211d32028af107e548)

it is. 
my test: compiled this supersling, replaced only this binary on my device with Chimera which has Zebra 1.1~beta2-1 installed, runs as expected.
(and works also on my devices with Electra on iOS11.4.1 and unc0ver/checkra1n on iOS13.3)
